File: codegenerator/types/BBoolean.c
#include <stdio.h>
#include <stdlib.h>

typedef struct BBoolean {
	int value;
} BBoolean;

BBoolean create_boolean(int value) {
    BBoolean result = {value};
    return result;
}

int booleanValue(BBoolean b) {
    return b.value;
}


BBoolean or(BBoolean b1, BBoolean b2) {
    return create_boolean(b1.value || b2.value);
}

BBoolean and(BBoolean b1, BBoolean b2) {
    return create_boolean(b1.value && b2.value);
}

BBoolean xor(BBoolean b1, BBoolean b2) {
    return create_boolean(b1.value != b2.value);
}

int main() {

    BBoolean a = create_boolean(1);
    BBoolean b = create_boolean(0);

    BBoolean c = or(a,b);
    BBoolean d = and(a,b);
    BBoolean e = xor(a,b);

    printf("%d\n", c.value);
    printf("%d\n", d.value);
    printf("%d\n", e.value);


}

File: codegenerator/types/BBoolean.java
package de.bmoth.codegenerator.types;

public class BBoolean implements BObject {
	private final Boolean value;

	public static boolean parseBoolean(String s) {
		return Boolean.parseBoolean(s);
	}

	public static String toString(boolean b) {
		return Boolean.toString(b);
	}

	public static Boolean valueOf(boolean b) {
		return Boolean.valueOf(b);
	}

	public int compareTo(Boolean b) {
		return value.compareTo(b);
	}

	public static Boolean valueOf(String s) {
		return Boolean.valueOf(s);
	}

	public static int compare(boolean x, boolean y) {
		return Boolean.compare(x, y);
	}

	public static boolean getBoolean(String name) {
		return Boolean.getBoolean(name);
	}

	public boolean booleanValue() {
		return value.booleanValue();
	}

	@Override
	public String toString() {
		return value.toString();
	}

	@Override
	public int hashCode() {
		return value.hashCode();
	}

	@Override
	public boolean equals(Object obj) {
		return value.equals(obj);
	}

	public BBoolean(boolean value) {
		this.value = new Boolean(value);
	}

	public BBoolean(String s) {
		this.value = new Boolean(s);
	}

	/* groovy operator overloading support */
	@SuppressWarnings("rawtypes")
	Object asType(Class clazz) {
		if (clazz == new Boolean(true).getClass()) {
			return this.booleanValue();
		}
		return this;
	}

	public boolean or(BBoolean other) {
		return this.booleanValue() || other.booleanValue();
	}

	public boolean or(Boolean other) {
		return this.booleanValue() || other;
	}

	public boolean xor(BBoolean other) {
		return this.booleanValue() ^ other.booleanValue();
	}

	public boolean xor(Boolean other) {
		return this.booleanValue() ^ other;
	}

	public boolean and(BBoolean other) {
		return this.booleanValue() && other.booleanValue();
	}

	public boolean and(Boolean other) {
		return this.booleanValue() && other;
	}
}
